### PR TITLE
Redux: enable Terms in global state tree.

### PR DESCRIPTION
**Pull request by timmyc**
Small follow up to #5350.  This branch adds the terms reducer to the global state tree making it available for usage in #5366.  Additionally discovered a bug where the query object was not being passed along in `<QueryTerms />` nor in the `TERMS_RECEIVE` action.

__To Test__
Ensure the terms tests pass by running `npm run test-client -- --grep "state terms"`

/cc @aduth

### CODE REVIEW

File: client/state/terms/schema.js [hunk 1]
```diff
@@ -16,7 +16,8 @@ export const termsSchema = {
 								slug: { type: 'string' },
 								description: { type: 'string' },
 								post_count: { type: 'number' },
-								parent: { type: 'number' }
+								parent: { type: 'number' },
+								meta: { type: 'object' }
```
Comment (timmyc) [inline client/state/terms/schema.js]: Also was getting warnings about the missing `meta` option in this Object so quick fix for that here as well.

Comment (aduth) [inline client/state/terms/schema.js]: >Also was getting warnings about the missing `meta` option in this Object so quick fix for that here as well.

Hmm, I wouldn't expect warnings for properties that aren't part of the schema, cc @gwwar . Got an example?

File: client/state/terms/actions.js [hunk 2]
```diff
@@ -16,14 +16,16 @@ import {
...
  */
-export function receiveTerms( siteId, taxonomy, terms ) {
+export function receiveTerms( siteId, taxonomy, terms, query = {} ) {
 	return {
 		type: TERMS_RECEIVE,
 		siteId,
 		taxonomy,
-		terms
+		terms,
+		query
```
Comment (aduth) [inline client/state/terms/actions.js]: We track queries using the `POSTS_REQUEST_SUCCESS` action type in the posts subtree. I'm not sure I like that approach, but we should think about a consistent approach. I would prefer keeping `POSTS_REQUEST_SUCCESS` limited in its scope (simply to request tracking), but at the same time, I'm not sure "receive" should necessarily be tied to a query*. Here, I think we should either split off a separate action creator/type, or allow `query` to be null (in which case, the `queries` reducer would ignore received items for that action).

_*_ The counter-example being, if I come across a term from some other API endpoint, I should be able to issue an ad-hoc `receiveTerms`.

**Comment by aduth** [comment]
>Ensure the terms tests pass by running `npm run test-client -- --grep "state terms"`

As of #5465, you can now also run the directory of tests using `npm run test-client client/state/terms` . Maybe ever so slightly more convenient 😄

File: client/state/terms/schema.js (hunk 1 above)
Comment (timmyc) [inline client/state/terms/schema.js]: I believe the warning is due to `additionalProperties` being false here

File: client/state/terms/actions.js (hunk 2 above)
Comment (timmyc) [inline client/state/terms/actions.js]: Makes total sense.  The other route is to pass along the terms to `TERMS_REQUEST_SUCCESS` - would you be okay with that approach @aduth ?

File: client/state/terms/schema.js [hunk 3]
```diff
@@ -16,7 +16,8 @@ export const termsSchema = {
...
 							},
 							additionalProperties: false
```
Comment (gwwar) [inline client/state/terms/schema.js]: > Hmm, I wouldn't expect warnings for properties that aren't part of the schema
This one is one level too deep. We only need `additionalProperties: false` for the patternProperties matches

**Comment by timmyc** [comment]
@aduth introduced `TERMS_RECEIVE_FOR_QUERY` to handle the use case as discussed and removed the invalid additionalProperties usage.

**Comment by mtias** [comment]
Could you include this in the #5046 list, please?

**Comment by aduth** [comment]
I hate to do this, but the more I look at the new action creator, the more I think we should just include `query` as an _optional_ property on the `TERMS_RECEIVE` action.

I think the issues with the current approach are:
- One must dispatch separate actions for receiving terms and, subsequently, terms for a specific query
- We could change this to allow the `items` reducer to also handle `RECEIVE_TERMS_FOR_QUERY`, but then we'd need to do this for any/all reducers needing to receive terms
- Dilutes action types and action creators with specialized forms (I could see one making the argument that specific handlers are a good thing)
- The naming `TERMS_RECEIVE_FOR_QUERY` doesn't really fall in line with our [guideline to suffix verb in action type naming](https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#actions) (a minor point)

What would you think about falling back to the implementation in a2d16a4c61c9e84aa8e529a0ffbae235bbacc0f0, except not providing a default for the `query` argument in `receiveTerms`, and adding a condition to `break` in the `state.terms.queries` reducer if the action object doesn't include the `query` property? 

I'm also open to being convinced otherwise 😄  Sorry to keep derailing your PRs.

**Comment by timmyc** [comment]
> and adding a condition to break in the state.terms.queries reducer if the action object doesn't include the query property?

Don't you think we should have the ability to have an empty query in the query mapping?  That way in QueryTerms we can always use the one selector instead of determining if we have an empty ( default ) query?

**Comment by timmyc** [comment]
@aduth reverted back to the original implementation and squashed in the removal of the errant `additionalProperties`

**Comment by aduth** [comment]
>Don't you think we should have the ability to have an empty query in the query mapping?

If I understand this right, this should be what we expect, even without the default query value in `receiveTerms`, because `requestTerms` _does_ have a default empty `query` object, so when it calls `receiveTerms` after receiving the results, we know that `query` will never be `undefined`, and we still have the option to issue ad-hoc `receiveTerms` without it being necessarily being tied to a query.

**Comment by timmyc** [comment]
Indeed you are correct!  Updated in edf6a2b

**Comment by aduth** [comment]
Looks good! 🚢

**Pull request merged**